### ISSUE TITLE
LabRecorder C++ Edition - More detailed build instructions needed

### ISSUE THREAD

**Issue by cboulay** [sccn/labstreaminglayer#98]
I attempted to write build instructions that worked on my system, and I had to modify the visual studio solution a little. Eventually I got it to build but then I was faced with an error when trying to run it. Any ideas?
![labrecorder_error](https://cloud.githubusercontent.com/assets/303797/13613105/9de14df6-e538-11e5-965d-af140a6302bb.PNG)

One of the modifications, which should apply to everyone, is to change the few occurrences of ../../../Qt/Desktop to ${QtSDK}/Desktop.
The other modification was to keep in-line with the default directory structure used by the pre-compiled boost binaries.
[Here are my changes](https://github.com/cboulay/labstreaminglayer/commit/701fef7b277dfd3c4bcb735fae26b18d064edb89). (edited to link directly to the commit)

I guess it's a little arbitrary that I chose to conform to the non-standard Qt directory structure that you use but not to the boost directory structure that you use.

Either way, it's a little silly for me to download a different compiler, different version of Qt, different version of boost, different liblsl binaries (I only had 64-bit on my system), use non-default paths, and add 2 environment variables (that will conflict with my other projects) just so I can use the provided visual studio project. All of these things can be abstracted away by using cmake or qmake.

If I find the time then I'm willing to write a CMakeLists.txt for this target, but I'd like to get the original building and running correctly first, if possible, so I have a good baseline.

**Comment by dmedine** [comment]
Hmmm, I guess I wasn't very careful about making the build system 
generic. Ideally the VS project should build on any machine provided the 
Qt and Boost environment variables are locally set.

I have no idea why the run failed. Are the liblsl and qt library 
binaries in the directory? I realize now I should update get_deps.py to 
grab these things correctly now that the directory tree has changed.

I am also going to add a build script for OSX and Linux (as is the case 
for the pupilpro and xdfbrowser apps). In fact, it should be the case 
that all apps that don't depend on windows-only libs should have cross 
platform build scripts. I have been using the qt build chain to do this 
thus far, but maybe cmake is better/more generic. I personally find 
cmake kind of bizarre.

I will take a look at the VS project tomorrow and try to get it more 
universal (and runtime error free!).

**Comment by cboulay** [comment]
For the visual studio project, I recommend changing all instances of `$(QTSDK)\Desktop\Qt\4.8.1\msvc2008\` to `$(QTSDK)\`, then updating your local environment variable to reflect that. Then it becomes possible to use the default Qt folder structure (from the precompiled binary downloads) and even use different versions of Qt built for different versions of MSVC.
$(BOOST_ROOT) seems to be OK as it is.
I'm guessing that these changes will enable an automatic update of the project and solution files to work correctly with newer MSVC.

I did have the lsl dll file in the same folder (after I first got an error about missing dlls), but I didn't put Qt dlls. Which are needed? Would it be better to link Qt statically?

**Comment by dmedine** [comment]
After looking at this more closely, I realize that these changes are a 
bit more involved than they appear at first blush. The reason the 
project is structured the way it is (which isn't necessarily a good 
reason) is that that is how all the other apps are structured. In fact, 
I created that project by copying the Mouse project and changing the 
names and code. I wanted to keep the references to outside libs uniform.

However, (and this is probably a different thread) it may in fact be 
time to revamp these things. MSVC 2008 is getting a bit old, but it is 
good to build this way for back compatibility. Another issue is that 
MSVC 2008 projects do not open in newer versions of Visual Studio -- 
which (@Microsoft) is incredibly annoying. Again, whether or not it is 
time to update all the projects for newer versions of VS (and Qt) is 
something that we ought to discuss, but for the time being, I'm inclined 
to leave things as they are.

I will make some notes about installing the Qt SDK and setting up the 
environment variables correctly to make things smoother for people that 
want to compile these projects on themselves.

Also, I am definitely opposed to maintaining different versions of the 
build environment for different versions of VS and Qt. .

**Comment by cboulay** [comment]
>Also, I am definitely opposed to maintaining different versions of the build environment for different versions of VS and Qt.

I'm not sure if that's in reference to something I said or not. Anyway, what I meant to convey is that by removing the version-specific folder names from the project file then you only have to maintain 1 build system but it will be compatible with other versions of VS, Qt, and Boost as long as the user has the environment variables set correctly and as long as the different versions of the software don't introduce incompatibilities. For most of the Apps, this shouldn't be a problem.

From reading your concerns, I think you will benefit greatly from using CMake or qmake. When it's setup correctly it is trivial to generate project files for almost any version of MSVC or any other ide/compiler/linker on any platform. Getting it set correctly can be difficult the first time you do it, but this can probably be aided greatly by a couple less well-known features of CMake.

You won't encounter these in many general tutorials, but take a look at [external projects](https://cmake.org/cmake/help/latest/module/ExternalProject.html), and [bundle utilities](https://cmake.org/cmake/help/latest/module/BundleUtilities.html) (not just for OS X app bundles).

I wish I had the time to tackle this but I definitely don't right now and not for several weeks at least.

**Comment by dmedine** [comment]
I'm not sure if that's in reference to something I said or not.

Nope. Just thinking out loud.

So far I'm not convinced that moving all 33 LSL Apps to CMake is 
necessarily the correct thing to do. As I see it, using CMake (or qmake) 
is a bit like robbing Peter to pay Paul. While it may automate the 
generation of build scripts and project files, it demands that the 
developer be familiar with both CMake and whatever environment CMake is 
targeting. It will not perform perfectly on every system and tweaks are 
inevitably necessary. It is not probably unreasonable to assume that a 
developer can do this, but neither is expecting her to be able to make 
minor changes to environment variables and linker properties in Visual 
Studio.

The other thing is that Visual Studio is the standard development tool 
for Windows and most of the LSL apps interface with hardware that is 
only supported on Windows. Thus far it has been less work for me to make 
build scripts (which lean on qmake, incidentally) for the few apps that 
will even run on Linux and OSX (so far, only PupilPro and XDFBrowser). 
To me, this is easier than dealing with CMake, but it is also less robust.

I still need to do this for the new version of LabRecorder.

That being said, if you feel strongly about CMake, I am definitely open 
to it! My feelings about it are just one man's opinion and in the end I 
don't really care what we do, provided it works.

For the time being, though, this is a moot point because I am also far 
too busy to deal with it. I propose that we close this issue and revisit 
the discussion when we are in a better position to actually do something 
about it.

**Comment by cboulay** [comment]
Sorry, I meant to answer last week. I was able to build and run on my PC. It was that the Qt dlls were missing, but the error didn't pop-up until I ran it in Debug mode.

Here are some instructions. I can put them into a pull request or you can modify them and add them to the LabRecorder/readme.txt .

```
Requirements:

labstreaminglayer:
This will only build within the full labstreaminglayer repository including the binaries typically obtained with `python get_deps.py`

Visual Studio:
Download and install [Visual Studio 2008](http://go.microsoft.com/?linkid=7729279)

Qt:
Download and install [QtSDK 4.8.1 for MSVC2008](https://download.qt.io/archive/qt/4.8/4.8.1/qt-win-opensource-4.8.1-vs2008.exe)
Using the default install folder for Qt will put it in C:\Qt\4.8.1.
The provided LabRecorder project assumes a different directory structure.
Create a folder at C:\Qt\Desktop\Qt\4.8.1\msvc2008
Move the contents of C:\Qt\4.8.1 to C:\Qt\Desktop\Qt\4.8.1\msvc2008
Add an environment variable QTSDK set to C:\Qt

Boost:
Download and install [Boost 1.59 for MSVC2008 32-bit](https://sourceforge.net/projects/boost/files/boost-binaries/1.59.0/boost_1_59_0-msvc-9.0-32.exe/download)
Add an environment variable BOOST_ROOT set to the boost install directory (C:\local\boost_1_59_0 by default)
Rename ${BOOST_ROOT}\lib32-msvc-9.0 to ${BOOST_ROOT}\lib

Building:
If, like me, you also have newer versions of MSVC installed
then it might be impossible to load the project by double-clicking on LabRecorder.sln
Open MSVC2008 from C:\Program Files (x86)\Microsoft Visual Studio 9.0\Common7\IDE\VCExpress.exe
File->Open->Project/Solution... then navigate to labstreaminglayer\Apps\LabRecorder\LabRecorder.sln
At the top, change from "Debug" to "Release".
Under the Build menu, click "Build Solution" (F7 shortcut key)

Running:
You will need liblsl32.dll, QtCore4.dll, and QtGui4.dll in the same folder as the executable.
```

**Comment by dmedine** [comment]
The instructions look very clear to me. I think we should definitely add 
this to the file APP BUILD ENVIRONMENT,txt

I think I am starting to understand what is going on with the Qt 
structure. It was so long ago that I downloaded and installed the SDK 
that I have no idea what actually went on, but I am sure that I didn't 
redo any directory structure. I can only assume that later editions of 
the SDK were structured differently (i.e. without the Qt\Destkop\Qt\etc. 
directories). Nevertheless, it is not just LabRecorder but all the apps 
whose properties settings assume this structure.

I never know what the right thing to do in a situation like this is. We 
can either:
1.) tell people to make the directory structure that adheres to the 
properties in the solutions to all the apps, or
2.) change the properties so that it adheres to Qt's newer(?) SDK.

If we go with 1.) we can use Chadwick's instructions as they are. I 
would just change the line about LabRecorder since AFAIK it is all the 
apps that are structured this way, and maybe make a note that some 
distributions of Qt's SDK already have this structure.

As I write I am starting to think that we should just do this for now 
since when MSVC2008 is finally dead, we will have to redo all the 
solutions anyway. No need to do it twice.

**Comment by dmedine** [comment]
I'm sorry to report that this problem is much worse than it appears. Part of the problem is that I don't understand VS2008 and part of the reason for that is that it makes absolutely no sense whatsoever. I think what has happened is that I hacked Qt into the project files/my own directory structure, then forgot what I did, then had to do that again, etc. etc. So, now there is mess and I don't know how to clean it up. Somehow these weird environment variables and command line options have spread like a virus (thanks, Microsoft!) and it is entirely opaque to me how to fix it. I will revisit this as soon as I can, because this stuff is terrible.

**Comment by dmedine** [comment]
Ok, so I figured out how to force vs2008 to bend to  my will. In the latest push to labrecorder, I incorporated the changes to the vcproj file so that there is no need for this \Desktop\etc. business. Now that I know how to do this, I can write a python script to sift through all the projects in the apps folder and make the appropriate corrections. Once this is done and the builds are tested, I will incorporate Chadwicks instructions (without the bit about moving the Qt directory tree around) and all will be well with the world. I'm going to leave this issue open until that is done.

**Comment by cboulay** [comment]
I just tried building but it failed in both Debug and Release. Here is the output of the Debug build attempt.

I'm actually running an experiment today. I do have a few streams that I'd rather not store so it would be great if I could get this working today but I can use the old one for today if this isn't an easy fix.

```
1>------ Build started: Project: LabRecorder, Configuration: Debug Win32 ------
1>Compiling...
1>mainwindow.cpp
1>C:\boost_1_59_0\boost/thread/future.hpp(410) : warning C4100: 'lk' : unreferenced formal parameter
1>e:\sachslab\tools\neurophys\labstreaminglayer\apps\labrecorder\recording.h(645) : warning C4100: 'phase_locked' : unreferenced formal parameter
1>.\mainwindow.cpp(85) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(85) : error C2228: left of '.c_str' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(142) : warning C4018: '<' : signed/unsigned mismatch
1>.\mainwindow.cpp(155) : warning C4018: '<' : signed/unsigned mismatch
1>.\mainwindow.cpp(170) : warning C4129: '%' : unrecognized character escape sequence
1>.\mainwindow.cpp(171) : warning C4129: '%' : unrecognized character escape sequence
1>.\mainwindow.cpp(253) : warning C4018: '<' : signed/unsigned mismatch
1>.\mainwindow.cpp(323) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(324) : warning C4129: '%' : unrecognized character escape sequence
1>.\mainwindow.cpp(326) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(326) : error C2228: left of '.find' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(328) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(328) : error C2228: left of '.size' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(330) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(330) : error C2228: left of '.replace' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(333) : warning C4129: '%' : unrecognized character escape sequence
1>.\mainwindow.cpp(335) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(335) : error C2228: left of '.find' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(336) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(336) : error C2228: left of '.size' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(337) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(337) : error C2228: left of '.replace' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(342) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(342) : error C2228: left of '.c_str' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(343) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(343) : error C2228: left of '.find_last_of' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(346) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(346) : error C2228: left of '.substr' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(347) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(347) : error C2228: left of '.substr' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(347) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(347) : error C2228: left of '.size' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(352) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(352) : error C2228: left of '.c_str' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(355) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(355) : error C2228: left of '.c_str' must have class/struct/union
1>.\mainwindow.cpp(367) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(367) : error C2228: left of '.substr' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(367) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(367) : error C2228: left of '.find_last_of' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(380) : error C2065: 'checkedStreams' : undeclared identifier
1>.\mainwindow.cpp(380) : error C2228: left of '.clear' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(388) : error C2065: 'checkedStreams' : undeclared identifier
1>.\mainwindow.cpp(388) : error C2228: left of '.push_back' must have class/struct/union
1>        type is ''unknown-type''
1>.\mainwindow.cpp(398) : error C2065: 'recFilename' : undeclared identifier
1>.\mainwindow.cpp(398) : error C2065: 'checkedStreams' : undeclared identifier
1>main.cpp
1>C:\boost_1_59_0\boost/thread/future.hpp(410) : warning C4100: 'lk' : unreferenced formal parameter
1>e:\sachslab\tools\neurophys\labstreaminglayer\apps\labrecorder\recording.h(645) : warning C4100: 'phase_locked' : unreferenced formal parameter
1>Generating Code...
1>Build log was saved at "file://e:\SachsLab\Tools\Neurophys\labstreaminglayer\Apps\LabRecorder\debug\BuildLog.htm"
1>LabRecorder - 39 error(s), 11 warning(s)
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by dmedine** [comment]
I apologize. Apparently my changes to the header file mainwindow.h 
didn't make it into the last commit. I just pushed these. It should be 
there now.

The latest version of labrecorder compiled for Windows 7+ is also on our 
ftp:
ftp://sccn.ucsd.edu/pub/software/LSL/Apps/LabRecorder-1.12b.zip

**Comment by cboulay** [comment]
Thanks for the quick fix. I'll bookmark the ftp so I don't forget that next time.
It ended up not mattering because the patient wasn't suitable for research overall. Oh well. I have it for next time.

**Comment by dmedine** [comment]
No worries. It took all of 30 seconds. 

I'm going to leave this open because I still haven't completed cleaning up the build environments. The references to Qt are now as they ought to be, but I'm not sure this is true of Boost as well. I have to set up a clean system in order to be sure everything is as it should be.

**Comment by Cookdj0128** [comment]
Don't know where else to post this - but I have a issue, and this topic seems to be the closest match. I downloaded the lsl github repository onto my windows 7, 64bit machine. I'm running visual studio 14, 2015. I downloaded cmake as well. I followed the install instructions included in the github path labstreaminglayer/LSL/liblsl/INSTALL. I have photos of the output of the cmake step and the VS step here: https://goo.gl/nH7t3T https://goo.gl/TfRZ69. I also installed QT. My problem is that I'm still not seeing any .exe files in the labstreaminglayer folder on my local home directory. What am I missing? The youtube video made it look so easy!

**Comment by cboulay** [comment]
LSL isn't an .exe, it's a '.dll'. You built it without error. It's up to you to create an App that uses the dll file.

Or, you can build one of the already-existing Apps. Is there an existing App in particular that you want to use?

**Comment by Cookdj0128** [comment]
Yes, thanks so much for the help! I would very much like to use the app that streams the Brainproducts / BrainAmpSeries. I'd also like to do online processing of the signal in python, and send feedback to participants via psychopy. I was thinking about using mushu, wyrm, and pyff for that, but other suggestions are welcome.

**Comment by chkothe** [comment]
In addition to building the apps from source, you can also download app
binaries from the download mirror on ftp:
ftp://sccn.ucsd.edu/pub/software/LSL/ (I'd say that's the quickest way to
get up and running)

**Comment by cboulay** [comment]
@dmedine will soon include some detailed build instructions for each of the provided apps. They're basically the same as the instructions I posted above on Mar 14, except the Qt part has changed.

Old, incorrect text:
```
The provided LabRecorder project assumes a different directory structure.
Create a folder at C:\Qt\Desktop\Qt\4.8.1\msvc2008
Move the contents of C:\Qt\4.8.1 to C:\Qt\Desktop\Qt\4.8.1\msvc2008
Add an environment variable QTSDK set to C:\Qt
```

New, correct text:
`Add an environment variable QTSDK set to C:\Qt\4.8.1`

(I am not on my Windows machine so I can't verify where Qt's default installation resides. Basically, you want to set the environment variable to the folder that is the parent of Qt's `include` folder.)

But the rest of the text should be correct. Instead of using the .vcproj file for LabRecorder, open the .vcproj for [BrainAmpSeries](https://github.com/sccn/labstreaminglayer/blob/master/Apps/BrainProducts/BrainAmpSeries/BrainAmpSeries.vcproj).

As for the Python processing, there aren't any ready-to-go Python apps that can do anything useful. [Qusp](http://qusp.io/projects/) has their [Neuropype](http://neuropype.io/) that is in closed beta. I wrote my own but it's not at a stage where I'm willing to share & support it. Of mushu, wyrm, and pyff, only wyrm would be of any use to you (you already have a source module to replace mushu, and you already know you want to use psychopy, which eliminates pyff). In wyrm, you'd replace [this line](https://github.com/bbci/wyrm/blob/master/examples/online_experiment.py#L88) with code that pulls data from an LSL inlet and formats it to work with wyrm, then you'd end the block in the `while True` loop by pushing your control signal to an LSL outlet. You can see examples of how to setup LSL inlets and outlets in Python [in this repository](https://github.com/sccn/labstreaminglayer/tree/master/LSL/liblsl-Python/examples).

Your psychopy app would pull data from the control signal LSL inlet and then display that on the screen somehow. I don't use psychopy so I don't know how you would set that up. I deliver my feedback through UnrealEngine (VR).

**Comment by dmedine** [comment]
I just pushed a modified APP BUILD ENVIRONMENT.txt file containing a modified set of the instructions cboulay suggested earlier in this post. I believe it is now correct and will work on any computer. The one thing that I am not %100 sure about is boost. On my computer, all I need to do is to set the BOOST_ROOT environment variable to the appropriate folder, and windows and vs magically do the rest. Some of the .vcproj files still have references to subdirectories of BOOST_ROOT to find headers and libraries, but I think this is superfluous. You can read finer details about this question here: [](http://boost.2283326.n4.nabble.com/linking-and-including-in-visual-studio-td4685702.html)

**Comment by Cookdj0128** [comment]
This is extraordinarily helpful! I'll return to the lab where I'm working at on Wednesday and give it a try.

**Comment by chkothe** [comment]
When you set BOOST_ROOT, the VS projects will find the headers without any
sub-directories in the project settings, but to use any of the dlls/libs,
you need to add the /lib subdirectory of $(BOOST_ROOT) to the library
include path, or at least that's how it used to be. One thing that has
changed since then is that nowadays the 64-bit boost libs are in /lib64,
and maybe the boost folks have added some #pragma's to their code to make
VS find libs automatically in those sub-directories - should be
straightforward to check. Now I trink all LSL apps are build for 32-bit, so
the lib64 thing shouldn't be an issue...

**Comment by dmedine** [comment]
It's the darndest thing, but I am able to link without putting anything in the 'additional libraries search path' field. How can I verify this? 

As I say, though, I should make sure that each project has this field labeled correctly, just to be sure.

**Comment by cboulay** [comment]
Closed with #209
